src/main/java/sparksoniq/jsoniq/runtime/iterator/functions/base/Functions.java

```diff
@@ -396,19 +499,41 @@ public static boolean checkBuiltInFunctionExists(FunctionIdentifier identifier)
 
     public static RuntimeIterator getBuiltInFunctionIterator(
             FunctionIdentifier identifier,
-            IteratorMetadata metadata,
-            List<RuntimeIterator> arguments
+            List<RuntimeIterator> arguments,
+            IteratorMetadata metadata
     ) {
-        Class<? extends RuntimeIterator> functionClass = builtInFunctions.get(identifier);
-        try {
-            Constructor<? extends RuntimeIterator> ctor = functionClass.getConstructor(
-                List.class,
-                IteratorMetadata.class
+        BuiltinFunction builtinFunction = builtInFunctions.get(identifier);
+
+        for (int i = 0; i < arguments.size(); i++) {
+            arguments.set(
+                i,
+                new TypePromotionIterator(
```

Comment:
After disentangling (see my other comment), this can become:

new TreatAsIterator(
...
   new TypePromotionIterator(
        ...
        arguments.get(i),
    ...
)

where the iterators cleanly call each other without even being aware of each other, and without any encapsulation breach.